cocos/ui/UIRichText.cpp

```diff
@@ -1492,115 +1493,151 @@ void RichText::formatText()
-    if (!isWrappable(text))
+    bool isWrappable(const StringUtils::StringUTF8& text)
     {
-        if (startingNewLine)
-            return (int) text.length();
-        return 0;
+        const StringUtils::StringUTF8::CharUTF8Store& str = text.getString();
+        return std::any_of(str.begin(), str.end(), isUTF8CharWrappable);
     }
 
-    for(int idx = (int)text.size()-1; idx >=0; )
+    int findSplitPositionForWord(Label* label, const StringUtils::StringUTF8& text, float leftSpaceWidth, float newLineWidth)
     {
-        int newidx = getPrevWord(text, idx);
-        if (newidx >=0)
+        float textRendererWidth = label->getContentSize().width;
+        float originalLeftSpaceWidth = leftSpaceWidth + textRendererWidth;
+
+        bool startingNewLine = (newLineWidth == originalLeftSpaceWidth);
+        if (!isWrappable(text))
+            return (startingNewLine ? static_cast<int>(text.length()) : 0);
+
+        float overstepPercent = (-leftSpaceWidth) / textRendererWidth;
+        size_t stringLength = text.length();
+
+        // rough estimate
+        int leftLength = stringLength * (1.0f - overstepPercent);
+
+        // The adjustment of the new line position
+        int idx = getNextWordPos(text, leftLength);
+        std::string leftStr = text.getAsCharSequence(0, idx);
+        label->setString(leftStr);
+        textRendererWidth = label->getContentSize().width;
+        if (originalLeftSpaceWidth < textRendererWidth)  // Have protruding
         {
-            idx = newidx;
-            auto leftStr = Helper::getSubStringOfUTF8String(text, 0, idx);
-            label->setString(leftStr);
-            if (label->getContentSize().width <= originalLeftSpaceWidth)
-                return idx;
+            while (1)
+            {
+                // try to erase a word
+                int newidx = getPrevWordPos(text, idx);
+                if (newidx >= 0)
+                {
+                    leftStr = text.getAsCharSequence(0, newidx);
+                    label->setString(leftStr);
+                    textRendererWidth = label->getContentSize().width;
+                    if (textRendererWidth <= originalLeftSpaceWidth)  // is fitted
+                        return newidx;
+                    idx = newidx;
+                    continue;
+                }
+                // newidx < 0 means no prev word
+                return (startingNewLine ? idx : 0);
+            }
         }
-        else
+        else if (textRendererWidth < originalLeftSpaceWidth)  // A wide margin
         {
-            if (startingNewLine)
+            while (1)
+            {
+                // try to append a word
+                int newidx = getNextWordPos(text, idx);
+                leftStr = text.getAsCharSequence(0, newidx);
+                label->setString(leftStr);
+                textRendererWidth = label->getContentSize().width;
+                if (textRendererWidth < originalLeftSpaceWidth)
```

Comment:
if textRendererWidth == originalLeftSpaceWidth, then should return `newidx` instead of `idx`

Reply:
Yes. I will add an if condition